torch/quantization/observer.py

```diff
@@ -1,50 +1,62 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
-import torch.nn as nn
-import torch
+
+from abc import ABC, abstractmethod
 from functools import partial
 
-class Observer(nn.Module):
-    r"""Default Observer Module
-    A default implementation of the observer module, only works for
-    `per_tensor_affine` quantization scheme.
-    The module will record the running average of max and min value of the
-    observed Tensor and calulate_qparams will calculate the scale and zero_point
-
-    Other types of Observers should follow the same API, it can take arbitrary
-    number of keyward arguments. In forward, it will update the statistics of
-    the observed Tensor. And it should provide a `calculate_qparam` function
-    that computes the quantization parameters given the collected statistics.
-    TODO: Maybe add an abstract Observer class that enforces these rules?
+import torch
+import torch.nn as nn
+
+
+class ObserverBase(ABC, nn.Module):
+    r"""Observer base Module
+    Any concrete observer implementation should derive from this class.
+
+    Concrete observers should follow the same API. In forward, they will update
+    the statistics of the observed Tensor. And they should provide a
+    `calculate_qparams` function that computes the quantization parameters given
+    the collected statistics.
     """
+
     def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine):
-        super(Observer, self).__init__()
+        super(ObserverBase, self).__init__()
         self.dtype = dtype
         self.qscheme = qscheme
-        assert self.qscheme in (torch.per_tensor_affine, torch.per_tensor_symmetric), \
-            'Default Observer only works for per_tensor_affine and \
-                per_tensor_symmetric quantization scheme'
-        assert self.dtype in (torch.qint8, torch.quint8), \
-            'Default Observer only works for qint8 and quint data type'
-        self.min_val = None
-        self.max_val = None
+        assert self.qscheme in (
+            torch.per_tensor_affine,
+            torch.per_tensor_symmetric,
```

Comment:
we don't have much support for per_tensor_symmetric I think

Reply:
Lets continue to support it. This only affects calculate q params

Reply:
hmm, theoretically we need to pass on the qscheme to quantized tensor, it works only because we always call `quantize_linear` when we convert quantized modules right now. Or should we always represent per_tensor_symmetric using per_tensor_affine quantization in backend?

Reply:
We need a way for backends to know that zero-point is zero (for uint8) and 128( for int8). Perhaps use the same backend and have a field to state that zero-point maps to zero/128?

Reply:
A n00b question, what's the difference between per_tensor_affine and per_tensor_symmetric here?

Reply:
@raghuramank100 what do you mean? if it is symmetric quantization, don't we have separate implementation of quantized kernels?

@hx89 per_tensor_symmetric only has scale as an argument and the float_min/float_max are always symmetric, please see: https://arxiv.org/pdf/1806.08342.pdf